library/kani/src/aliasing.rs

```diff
@@ -0,0 +1,313 @@
+    /// Disabled -- no accesses allowed
+    const DISABLED: u8 = 3;
+}
+
+impl Permission {
+    /// Returns whether the access bit is granted by the permission
+    /// byte
+    fn grants(access: AccessBit, perm: PermissionByte) -> bool {
+        perm != Permission::DISABLED && (access != Access::WRITE || perm != Permission::SHAREDRO)
+    }
+}
+
+/// Associate every pointer object with a permission
+static mut PERMS: ShadowMem<PermissionByte> = ShadowMem::new(Permission::SHAREDRO);
+
+/// State of the borrows stack monitor for a byte
+pub(super) mod monitors {
+    /// Tracks whether the monitor is on or off.
+    /// Encoded as associated constants instead
+    /// of as an enum to ensure that the representation
+    /// uses 1 bit.
+    type MonitorBit = bool;
+    struct MonitorState;
+    impl MonitorState {
+        const ON: MonitorBit = false;
+        const OFF: MonitorBit = true;
+    }
+
+    #[allow(unused)]
+    const STACK_DEPTH: usize = 15;
+
+    /// Whether the monitor is on. Initially, the monitor is
+    /// "off", and it will remain so until an allocation is found
+    /// to track.
+    static mut STATE: MonitorBit = MonitorState::OFF;
+    /// Object + offset being monitored
+    pub static mut MONITORED: *const u8 = std::ptr::null();
+    /// The tags of the pointer objects borrowing the byte
+    static mut STACK_TAGS: [PointerTag; STACK_DEPTH] = [0; STACK_DEPTH];
+    /// The permissions of the pointer objects borrowing the byte
+    static mut STACK_PERMS: [PermissionByte; STACK_DEPTH] = [Permission::UNIQUE; STACK_DEPTH];
+    /// The "top" of the stack
+    static mut STACK_TOP: usize = 0;
+
+    use super::*;
+
+    /// Initialize local when track local is true, picking a monitor,
+    /// and setting its object and offset to within pointer.
+    pub(super) unsafe fn track_local<U>(tag: PointerTag, pointer: *const U) {
+        // Decide whether to initialize the stacks
+        // for location:location+size_of(U).
+        unsafe {
+            if demonic_nondet() && STATE == MonitorState::OFF {
+                STATE = MonitorState::ON;
+                let offset: usize = kani::any();
+                crate::assume(offset < std::mem::size_of::<U>());
+                MONITORED = pointer.byte_add(offset) as *const u8;
+                STACK_TAGS[STACK_TOP] = tag;
+                STACK_PERMS[STACK_TOP] = Permission::UNIQUE;
+                STACK_TOP += 1;
+            }
+        }
+    }
+
+    /// Push a tag with a permission perm at pointer
+    pub(super) fn push<U>(tag: PointerTag, perm: PermissionByte, pointer: *const U) {
+        // Decide whether to initialize the stacks
+        // for location:location+size_of(U).
+        // Offset has already been picked earlier.
+        unsafe {
+            use self::*;
+            if STATE == MonitorState::ON
+                && pointer_object(MONITORED) == pointer_object(pointer)
+                && pointer_offset(MONITORED) <= std::mem::size_of::<U>()
```

Comment:
I don't think this will work for a struct field or an array element. This will also not work for unsized types (I believe you need to use `size_of_val_raw` for that and add `U: ?Sized`). If you do, please make this function `unsafe`.

I think you need this to be something like:
```suggestion
                && MONITORED.addr() >= pointer.addr()
                && MONITORED.addr() < pointer.addr().wrapping_add(std::mem::size_of_val_raw(pointer))
```